registry/auth.go

```diff
@@ -169,28 +181,49 @@ func (err PingResponseError) Error() string {
 // PingV2Registry attempts to ping a v2 registry and on success return a
 // challenge manager for the supported authentication types.
 // If a response is received but cannot be interpreted, a PingResponseError will be returned.
-func PingV2Registry(endpoint *url.URL, transport http.RoundTripper) (challenge.Manager, error) {
+func PingV2Registry(endpoint *url.URL, transport http.RoundTripper) (challenge.Manager, string, error) {
 	pingClient := &http.Client{
 		Transport: transport,
 		Timeout:   15 * time.Second,
 	}
 	endpointStr := strings.TrimRight(endpoint.String(), "/") + "/v2/"
 	req, err := http.NewRequest(http.MethodGet, endpointStr, nil)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	resp, err := pingClient.Do(req)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	defer resp.Body.Close()
 
 	challengeManager := challenge.NewSimpleManager()
 	if err := challengeManager.AddResponse(resp); err != nil {
-		return nil, PingResponseError{
+		return nil, "", PingResponseError{
 			Err: err,
 		}
 	}
 
-	return challengeManager, nil
+	responseUrl := url.URL{
+		Path:   resp.Request.URL.Path,
+		Host:   resp.Request.URL.Host,
+		Scheme: resp.Request.URL.Scheme,
+	}
+
+	challenges, err := challengeManager.GetChallenges(responseUrl)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if len(challenges) == 1 {
```

Comment:
Also, I see that the request will made for each challenge if there are more than one:
https://github.com/distribution/distribution/blob/c9b844e533fc3fb8626355bbfe013e3001e7d3fe/registry/client/auth/session.go#L97

So this will still fail later and perhaps we should log a warning.
How common would it be in a "real world" to have more than one challenge?

Reply:
I'm agree with you, then I added some TODO and comments in the code to support multiple challenges when loading TLS configs. I also added an warning when encountering a multiple challenges case.
I think the case I fixed represents the most common case in the real world. Generally when the client certificates are needed, these client certificates are the same for the docker registry and for the token provider (otherwise the problem I fixed should have been reported many times...). We introduced some advanced security practices that need different client certificates to access to a docker registry and its token provider, but only with one auth challenge => it represents a real win if the docker engine manage this case. It could manage more complicated cases in the future if needed, but I think as you said that it will probably be an API breakage